File: compiler/ast.cpp
#include "ast.hpp"
#include <iostream>
#include <sstream>

void ProgramNode::generate(const CodeContext& context) const {
    if (function_declaration) {
        if (auto* funcDef = dynamic_cast<FunctionDefinitionNode*>(function_declaration.get())) {
            funcDef->generate(context);
        }
    }
}

void FunctionDefinitionNode::generate(const CodeContext& context) {
	VariableResolutionVisitor resolver;
	accept(resolver);

    FunctionBody body{ identifier };

    TacVisitor visitor(body);
    accept(visitor);
    if (!dynamic_cast<ReturnInstruction*>(body.instructions.back().get()) && body.name == "main") { // Default return statement in main method
        body.emplaceInstruction<ReturnInstruction>(static_cast<unsigned int>(0));
    }

    std::stringstream ss;
    for (const auto& instruction : body.instructions) {
        instruction->makeAssembly(ss, body);
    }
    context.out << ss.str();
    std::cout << body << std::endl;
}

PrintVisitor::PrintVisitor(std::ostream& os, int indent)
    : os(os), indent(indent) {
}

void PrintVisitor::increaseIndent() {
    indent++;
}

void PrintVisitor::decreaseIndent() {
    if (indent > 0) {
	    indent--;
    }
}

std::string PrintVisitor::getIndent() const {
    return std::string(indent, ' ');
}

void PrintVisitor::visitProgram(ProgramNode* const node) {
    os << "PROGRAM NODE\n";
    increaseIndent();
    if (node->function_declaration) {
        node->function_declaration->accept(*this);
    }
    decreaseIndent();
}

void PrintVisitor::visitFunctionDefinition(FunctionDefinitionNode* const node) {
    os << getIndent() << "FUNCTION DECLARATION NODE: " << node->identifier << '\n';
    increaseIndent();
    for (const auto& statement : node->block_items) {
        statement->accept(*this);
    }
    decreaseIndent();
}

void PrintVisitor::visitDeclaration(DeclarationNode* const node) {
    os << getIndent() << "DECLARATION NODE: " << node->identifier << '\n';
    if (node->expression) {
        increaseIndent();
        node->expression->accept(*this);
        decreaseIndent();
    }
}

void PrintVisitor::visitAssignment(AssignmentNode* const node) {
    os << getIndent() << "ASSIGNMENT NODE:\n";
    increaseIndent();
    node->left->accept(*this);
    node->right->accept(*this);
    decreaseIndent();
}

void PrintVisitor::visitReturn(ReturnNode* const node) {
    os << getIndent() << "RETURN NODE\n";
    if (node->expression) {
        increaseIndent();
        node->expression->accept(*this);
        decreaseIndent();
    }
}

void PrintVisitor::visitUnary(UnaryNode* const node) {
    os << getIndent() << "UNARY NODE: ";
    switch (node->op) {
    case UnaryOperator::NEGATION:
        os << "MINUS\n";
        break;
    case UnaryOperator::BITWISE_NOT:
        os << "BITWISE NOT\n";
        break;
    case UnaryOperator::LOGICAL_NOT:
        os << "LOGICAL NOT\n";
        break;
    }
    increaseIndent();
    node->expression->accept(*this);
    decreaseIndent();
}

void PrintVisitor::visitBinary(BinaryNode* const node) {
    os << getIndent();
    switch (node->op) {
    case BinaryOperator::ADD:
        os << "ADD\n";
        break;
    case BinaryOperator::SUBTRACT:
        os << "SUBTRACT\n";
        break;
    case BinaryOperator::MULTIPLY:
        os << "MULTIPLY\n";
        break;
    case BinaryOperator::DIVIDE:
        os << "DIVIDE\n";
        break;
        // Add other operators as needed
    }
    increaseIndent();
    node->left->accept(*this);
    node->right->accept(*this);
    decreaseIndent();
}

void PrintVisitor::visitConst(ConstNode* const node) {
    os << getIndent() << "CONST NODE: " << node->value << '\n';
}

void PrintVisitor::visitVariable(VariableNode* const node) {
    os << getIndent() << "VARIABLE NODE: " << node->identifier << '\n';
}

TacVisitor::TacVisitor(FunctionBody& body)
    : body(body), result(nullptr) {
}

Operand TacVisitor::getResult() const {
    return result;
}

void TacVisitor::visitProgram(ProgramNode* const node) {
    if (node->function_declaration) {
        node->function_declaration->accept(*this);
    }
}

void TacVisitor::visitFunctionDefinition(FunctionDefinitionNode* const node) {
    body.emplaceInstruction<FunctionInstruction>(body.name);
    body.emplaceInstruction<AllocateStackInstruction>();

    for (const auto& statement : node->block_items) {
        statement->accept(*this);
    }
}

void TacVisitor::visitDeclaration(DeclarationNode* const node) {
	PseudoRegister pseudoRegister{body.name, body.variableCount};
    body.variableToPseudoregister[node->identifier] = pseudoRegister;
    if (node->expression) {
        node->expression->accept(*this);
        body.emplaceInstruction<StoreValueInstruction>(result);
    }
	body.variableCount++;
}

void TacVisitor::visitAssignment(AssignmentNode* const node) {
	node->right->accept(*this);
    try {
        node->left->accept(*this);
        PseudoRegister variable = std::get<PseudoRegister>(result);
        node->right->accept(*this);
        Operand src = result;
        body.emplaceInstruction<StoreValueInstruction>(variable, src);
	} catch (std::bad_variant_access&) {
		throw semantic_error("Invalid lvalue!");
	}
}

void TacVisitor::visitReturn(ReturnNode* const node) {
    Operand dest = nullptr;
    if (node->expression) {
        node->expression->accept(*this);
        dest = result;
    }
    body.emplaceInstruction<ReturnInstruction>(dest);
    result = nullptr;
}

void TacVisitor::visitUnary(UnaryNode* const node) {
    node->expression->accept(*this);
    Operand src = result;

    PseudoRegister dest = body.emplaceInstruction<UnaryOpInstruction>(node->op, src);
    body.variableCount++;
    result = dest;
}

void TacVisitor::visitBinary(BinaryNode* const node) {
    if (node->op == BinaryOperator::LOGICAL_AND) {
        std::string falseLabel = std::format(".{}{}_false", body.name, ++body.labelCount);
        std::string endLabel = std::format(".{}{}_end", body.name, ++body.labelCount);

        // Short-circuiting
        node->left->accept(*this);
        Operand leftOperand = result;
        body.emplaceInstruction<JumpIfZero>(leftOperand, falseLabel); // goto false label

        node->right->accept(*this);
        Operand rightOperand = result;
        body.emplaceInstruction<JumpIfZero>(rightOperand, falseLabel);

        PseudoRegister dest = body.emplaceInstruction<StoreValueInstruction>(static_cast<Number>(1));
        body.emplaceInstruction<Jump>(endLabel); // goto end

        body.emplaceInstruction<Label>(falseLabel); // false label
        dest = body.emplaceInstruction<StoreValueInstruction>(static_cast<Number>(0));

        body.emplaceInstruction<Label>(endLabel); // end
        body.variableCount++;
        result = dest;
        return;
    }

    if (node->op == BinaryOperator::LOGICAL_OR) {
        std::string trueLabel = std::format(".{}{}_true", body.name, ++body.labelCount);
        std::string endLabel = std::format(".{}{}_end", body.name, ++body.labelCount);

        // Short-circuiting
        node->left->accept(*this);
        Operand leftOperand = result;
        body.emplaceInstruction<JumpIfNotZero>(leftOperand, trueLabel); // goto true label

        node->right->accept(*this);
        Operand rightOperand = result;
        body.emplaceInstruction<JumpIfNotZero>(rightOperand, trueLabel);

        PseudoRegister dest = body.emplaceInstruction<StoreValueInstruction>(static_cast<Number>(0));
        body.emplaceInstruction<Jump>(endLabel); // goto end

        body.emplaceInstruction<Label>(trueLabel); // true label
        dest = body.emplaceInstruction<StoreValueInstruction>(static_cast<Number>(1));

        body.emplaceInstruction<Label>(endLabel); // end
        body.variableCount++;
        result = dest;
        return;
    }

    node->left->accept(*this);
    Operand leftOperand = result;

    node->right->accept(*this);
    Operand rightOperand = result;

    PseudoRegister dest = body.emplaceInstruction<BinaryOpInstruction>(node->op, leftOperand, rightOperand);
    body.variableCount++;
    result = dest;
}

void TacVisitor::visitConst(ConstNode* const node) {
    result = node->value;
}

void TacVisitor::visitVariable(VariableNode* const node) {
	if (!body.variableToPseudoregister.contains(node->identifier)) {
		throw semantic_error(std::format("Undeclared variable {}", node->identifier));
	}
    result = body.variableToPseudoregister[node->identifier];
}

void VariableResolutionVisitor::visitProgram(ProgramNode* const node) {
    if (node->function_declaration) {
        node->function_declaration->accept(*this);
    }
}

void VariableResolutionVisitor::visitFunctionDefinition(FunctionDefinitionNode* const node) {
    for (const auto& statement : node->block_items) {
        statement->accept(*this);
    }
}

/*
 *resolve_declaration(Declaration(name, init), variable_map):
 1 if name is in variable_map:
 fail("Duplicate variable declaration!")
 unique_name = make_temporary()
 2 variable_map.add(name, unique_name)
 3 if init is not null:
 init = resolve_exp(init, variable_map)
 4 return Declaration(unique_name, init)
 */
void VariableResolutionVisitor::visitDeclaration(DeclarationNode* const node) {
    if (variableMap.contains(node->identifier)) {
        throw semantic_error(std::format("Duplicate variable declaration {}", node->identifier));
    }
    std::string newName = makeTemporary(node->identifier);
    variableMap[node->identifier] = newName;
	node->identifier = newName;

    if (node->expression) {
        node->expression->accept(*this);
    }
}

/*
 *resolve_exp(e, variable_map):
 match e with
 | Assignment(left, right) ->
 if left is not a Var node:
 fail("Invalid lvalue!")
 return Assignment(1 resolve_exp(left, variable_map), 2 resolve_exp(right, variable_map))
 | Var(v) ->
 if v is in variable_map:
 return Var(3 variable_map.get(v))
 else:
 fail("Undeclared variable!")
 | --snip--
 */
void VariableResolutionVisitor::visitAssignment(AssignmentNode* const node) {
	if (!dynamic_cast<VariableNode*>(node->left.get())) {
		throw semantic_error("Invalid lvalue!");
	}
	node->left->accept(*this);
	node->right->accept(*this);
}

void VariableResolutionVisitor::visitReturn(ReturnNode* const node) {
	if (node->expression) {
		node->expression->accept(*this);
	}
}

void VariableResolutionVisitor::visitUnary(UnaryNode* const node) {
	node->expression->accept(*this);
}

void VariableResolutionVisitor::visitBinary(BinaryNode* const node) {
	node->left->accept(*this);
	node->right->accept(*this);
}

void VariableResolutionVisitor::visitVariable(VariableNode* const node) {
	node->identifier = variableMap[node->identifier];
}

File: rust/src/ast.rs
use crate::common::Position;
use crate::errors::CompilerError;
use crate::lexer::{BinaryOperator, Number, UnaryOperator};
use crate::tac::FunctionBody;
use crate::tac_visitor::TacVisitor;
use crate::variable_resolution::VariableResolutionVisitor;
use std::rc::Rc;

pub trait Visitor {
    fn visit_program(
        &mut self,
        line_number: &Rc<Position>,
        function_declaration: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError>;
    fn visit_function(
        &mut self,
        line_number: &Rc<Position>,
        identifier: &mut Rc<String>,
        body: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError>;
    fn visit_declaration(
        &mut self,
        line_number: &Rc<Position>,
        identifier: &mut Rc<String>,
        expression: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError>;
    fn visit_assignment(
        &mut self,
        line_number: &Rc<Position>,
        left: &mut Box<ASTNode>,
        right: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError>;
    fn visit_return(
        &mut self,
        line_number: &Rc<Position>,
        expression: &mut Option<Box<ASTNode>>,
    ) -> Result<(), CompilerError>;
    fn visit_block(
        &mut self,
        line_number: &Rc<Position>,
        body: &mut Vec<Box<ASTNode>>,
    ) -> Result<(), CompilerError>;
    fn visit_unary(
        &mut self,
        line_number: &Rc<Position>,
        op: &mut UnaryOperator,
        expression: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError>;
    fn visit_binary(
        &mut self,
        line_number: &Rc<Position>,
        op: &mut BinaryOperator,
        left: &mut Box<ASTNode>,
        right: &mut Box<ASTNode>,
    ) -> Result<(), CompilerError>;
    fn visit_condition(
        &mut self,
        line_number: &Rc<Position>,
        condition: &mut Box<ASTNode>,
        if_true: &mut Option<Box<ASTNode>>,
        if_false: &mut Option<Box<ASTNode>>,
        is_ternary: &mut bool,
    ) -> Result<(), CompilerError>;
    fn visit_while(
        &mut self,
        line_number: &Rc<Position>,
        condition: &mut Box<ASTNode>,
        body: &mut Option<Box<ASTNode>>,
        label: &mut Rc<String>,
        is_do_while: &mut bool,
    ) -> Result<(), CompilerError>;
    fn visit_break(
        &mut self,
        line_number: &Rc<Position>,
        label: &mut Rc<String>,
    ) -> Result<(), CompilerError>;
    fn visit_continue(
        &mut self,
        line_number: &Rc<Position>,
        label: &mut Rc<String>,
        is_for: &mut bool,
    ) -> Result<(), CompilerError>;
    fn visit_for(
        &mut self,
        line_number: &Rc<Position>,
        init: &mut Option<Box<ASTNode>>,
        condition: &mut Option<Box<ASTNode>>,
        increment: &mut Option<Box<ASTNode>>,
        body: &mut Option<Box<ASTNode>>,
        label: &mut Rc<String>,
    ) -> Result<(), CompilerError>;
    fn visit_const(
        &mut self,
        line_number: &Rc<Position>,
        value: &mut Number,
    ) -> Result<(), CompilerError>;
    fn visit_variable(
        &mut self,
        line_number: &Rc<Position>,
        identifier: &mut Rc<String>,
    ) -> Result<(), CompilerError>;
    fn visit_prefix(
        &mut self,
        line_number: &Rc<Position>,
        variable: &mut Box<ASTNode>,
        operator: &mut UnaryOperator,
    ) -> Result<(), CompilerError>;
    fn visit_postfix(
        &mut self,
        line_number: &Rc<Position>,
        variable: &mut Box<ASTNode>,
        operator: &mut UnaryOperator,
    ) -> Result<(), CompilerError>;
}

#[derive(Debug)]
pub enum ASTNodeType {
    ProgramNode {
        function_declaration: Box<ASTNode>,
    },
    FunctionNode {
        identifier: Rc<String>,
        body: Option<Box<ASTNode>>,
    },
    BlockNode {
        body: Vec<Box<ASTNode>>,
    },
    DeclarationNode {
        identifier: Rc<String>,
        expression: Option<Box<ASTNode>>,
    },
    ReturnNode {
        expression: Option<Box<ASTNode>>,
    },
    UnaryNode {
        op: UnaryOperator,
        expression: Box<ASTNode>,
    },
    BinaryNode {
        op: BinaryOperator,
        left: Box<ASTNode>,
        right: Box<ASTNode>,
    },
    ConstNode {
        value: Number,
    },
    VariableNode {
        identifier: Rc<String>,
    },
    PrefixNode {
        variable: Box<ASTNode>,
        operator: UnaryOperator,
    },
    PostfixNode {
        variable: Box<ASTNode>,
        operator: UnaryOperator,
    },
    AssignmentNode {
        left: Box<ASTNode>,
        right: Box<ASTNode>,
    },
    ConditionNode {
        condition: Box<ASTNode>,
        if_true: Option<Box<ASTNode>>,
        if_false: Option<Box<ASTNode>>,
        is_ternary: bool,
    },
    WhileNode {
        condition: Box<ASTNode>,
        body: Option<Box<ASTNode>>,
        label: Rc<String>,
        is_do_while: bool,
    },
    BreakNode {
        label: Rc<String>,
    },
    ContinueNode {
        label: Rc<String>,
        is_for: bool,
    },
    ForNode {
        init: Option<Box<ASTNode>>,
        condition: Option<Box<ASTNode>>,
        increment: Option<Box<ASTNode>>,
        body: Option<Box<ASTNode>>,
        label: Rc<String>,
    },
}

#[derive(Debug)]
pub struct ASTNode {
    line_number: Rc<Position>,
    pub(crate) kind: ASTNodeType,
}

impl ASTNode {
    pub fn new(line_number: Rc<Position>, kind: ASTNodeType) -> ASTNode {
        ASTNode { line_number, kind }
    }

    pub fn accept(&mut self, visitor: &mut dyn Visitor) -> Result<(), CompilerError> {
        match &mut self.kind {
            ASTNodeType::ProgramNode {
                function_declaration,
            } => visitor.visit_program(&self.line_number, function_declaration),
            ASTNodeType::FunctionNode { identifier, body } => {
                visitor.visit_function(&self.line_number, identifier, body)
            }
            ASTNodeType::BlockNode { body } => visitor.visit_block(&self.line_number, body),
            ASTNodeType::DeclarationNode {
                identifier,
                expression,
            } => visitor.visit_declaration(&self.line_number, identifier, expression),
            ASTNodeType::ReturnNode { expression } => {
                visitor.visit_return(&self.line_number, expression)
            }
            ASTNodeType::UnaryNode { op, expression } => {
                visitor.visit_unary(&self.line_number, op, expression)
            }
            ASTNodeType::BinaryNode { op, left, right } => {
                visitor.visit_binary(&self.line_number, op, left, right)
            }
            ASTNodeType::ConditionNode {
                condition,
                if_true,
                if_false,
                is_ternary,
            } => {
                visitor.visit_condition(&self.line_number, condition, if_true, if_false, is_ternary)
            }
            ASTNodeType::WhileNode {
                condition,
                body,
                label,
                is_do_while,
            } => visitor.visit_while(&self.line_number, condition, body, label, is_do_while),
            ASTNodeType::BreakNode { label } => visitor.visit_break(&self.line_number, label),
            ASTNodeType::ContinueNode { label, is_for } => {
                visitor.visit_continue(&self.line_number, label, is_for)
            }
            ASTNodeType::ForNode {
                init,
                condition,
                increment,
                body,
                label,
            } => visitor.visit_for(&self.line_number, init, condition, increment, body, label),
            ASTNodeType::ConstNode { value } => visitor.visit_const(&self.line_number, value),
            ASTNodeType::VariableNode { identifier } => {
                visitor.visit_variable(&self.line_number, identifier)
            }
            ASTNodeType::PrefixNode { variable, operator } => {
                visitor.visit_prefix(&self.line_number, variable, operator)
            }
            ASTNodeType::AssignmentNode { left, right } => {
                visitor.visit_assignment(&self.line_number, left, right)
            }
            ASTNodeType::PostfixNode { variable, operator } => {
                visitor.visit_postfix(&self.line_number, variable, operator)
            }
        }
    }

    pub fn generate(&mut self, out: &mut String) -> Result<(), CompilerError> {
        match &self.kind {
            ASTNodeType::ProgramNode { .. } => {
                if let ASTNodeType::ProgramNode {
                    function_declaration,
                } = &mut self.kind
                {
                    function_declaration.generate(out)
                } else {
                    unreachable!()
                }
            }
            ASTNodeType::FunctionNode { .. } => generate_function_node(out, self),
            _ => unreachable!(),
        }
    }
}

fn generate_function_node(out: &mut String, this: &mut ASTNode) -> Result<(), CompilerError> {
    let identifier = match &this.kind {
        ASTNodeType::FunctionNode { identifier, body } => {
            if body.is_none() {
                return Ok(());
            }
            Rc::clone(&identifier)
        }
        _ => unreachable!(),
    };

    let mut variable_resolution_visitor = VariableResolutionVisitor::new(Rc::clone(&identifier));
    this.accept(&mut variable_resolution_visitor as &mut dyn Visitor)?;

    let mut function_body = FunctionBody::new();

    let mut tac_visitor = TacVisitor::new(Rc::clone(&identifier), &mut function_body);
    this.accept(&mut tac_visitor as &mut dyn Visitor)?;

    // Default return statement in the main method
    if identifier.as_str() == "main" {
        function_body.add_default_return_to_main(&this.line_number);
    }

    println!("{:?}", function_body);

    for instruction in &function_body.instructions {
        instruction.make_assembly(out, &function_body);
    }
    Ok(())
}

pub fn is_lvalue_node(node: &ASTNodeType) -> bool {
    match node {
        ASTNodeType::VariableNode { .. } | ASTNodeType::PrefixNode { .. } => true,
        _ => false,
    }
}

pub fn extract_base_variable(node: &ASTNodeType) -> Option<Rc<String>> {
    match node {
        ASTNodeType::VariableNode { identifier } => Some(identifier.clone()),
        ASTNodeType::PrefixNode { variable, .. } => extract_base_variable(&variable.kind),
        _ => None,
    }
}
